Add support for PickleOpCode::APPEND in torch unpickler

emasap opened the pull request:
Reviewed By: qiminglu

Differential Revision: D46760650

ezyang commented:
CI failure looks real:

```
2023-07-18T18:27:47.2804281Z ERROR: /var/lib/jenkins/workspace/BUILD.bazel:1617:11: Compiling torch/csrc/jit/serialization/unpickler.cpp failed: (Exit 1): gcc failed: error executing command (from target //:torch) /opt/cache/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++11' -MD -MF ... (remaining 323 arguments skipped)
2023-07-18T18:27:47.2805197Z 
2023-07-18T18:27:47.2805618Z Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
2023-07-18T18:27:47.2813709Z In file included from ./c10/util/Exception.h:4,
2023-07-18T18:27:47.2814552Z                  from aten/src/ATen/core/Generator.h:11,
2023-07-18T18:27:47.2814943Z                  from aten/src/ATen/CPUGeneratorImpl.h:3,
2023-07-18T18:27:47.2815324Z                  from aten/src/ATen/Context.h:3,
2023-07-18T18:27:47.2815661Z                  from aten/src/ATen/ATen.h:7,
2023-07-18T18:27:47.2816025Z                  from torch/csrc/jit/serialization/unpickler.cpp:1:
2023-07-18T18:27:47.2816954Z torch/csrc/jit/serialization/unpickler.cpp: In member function 'torch::jit::PickleOpCode torch::jit::Unpickler::readInstruction()':
2023-07-18T18:27:47.2958760Z torch/csrc/jit/serialization/unpickler.cpp:395:29: error: comparison of integer expressions of different signedness: 'size_t' {aka 'long unsigned int'} and 'long int' [-Werror=sign-compare]
2023-07-18T18:27:47.2959592Z   395 |           TORCH_CHECK(start <= std::numeric_limits<ssize_t>::max());
2023-07-18T18:27:47.2968374Z       |                       ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-07-18T18:27:47.2969495Z torch/csrc/jit/serialization/unpickler.cpp:473:25: error: comparison of integer expressions of different signedness: 'size_t' {aka 'long unsigned int'} and 'long int' [-Werror=sign-compare]
2023-07-18T18:27:47.2970163Z   473 |       TORCH_CHECK(start <= std::numeric_limits<ssize_t>::max());
2023-07-18T18:27:47.2970552Z       |                   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-07-18T18:27:47.2971448Z torch/csrc/jit/serialization/unpickler.cpp:495:25: error: comparison of integer expressions of different signedness: 'size_t' {aka 'long unsigned int'} and 'long int' [-Werror=sign-compare]
2023-07-18T18:27:47.2972107Z   495 |       TORCH_CHECK(start <= std::numeric_limits<ssize_t>::max());
2023-07-18T18:27:47.2972535Z       |                   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-07-18T18:27:47.2973699Z torch/csrc/jit/serialization/unpickler.cpp:636:33: error: comparison of integer expressions of different signedness: 'std::vector<c10::IValue>::size_type' {aka 'long unsigned int'} and 'long int' [-Werror=sign-compare]
2023-07-18T18:27:47.2974422Z   636 |       TORCH_CHECK(stack_.size() <= std::numeric_limits<ssize_t>::max());
2023-07-18T18:27:47.2982856Z       |                   ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-07-18T18:27:47.2983800Z torch/csrc/jit/serialization/unpickler.cpp: In member function 'void torch::jit::Unpickler::readListElements(c10::IValue, size_t)':
2023-07-18T18:27:47.2984561Z torch/csrc/jit/serialization/unpickler.cpp:1162:21: error: comparison of integer expressions of different signedness: 'size_t' {aka 'long unsigned int'} and 'long int' [-Werror=sign-compare]
2023-07-18T18:27:47.2984999Z  1162 |   TORCH_CHECK(start <= std::numeric_limits<ssize_t>::max());
2023-07-18T18:27:47.2985252Z       |               ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

lintrunner also wants you to clang-format your C++

```
2023-07-18T18:29:22.2865183Z     You can run `lintrunner -a` to apply this patch.
2023-07-18T18:29:22.2865377Z 
2023-07-18T18:29:22.2865563Z     451  451 |       readList(list_ivalue);
2023-07-18T18:29:22.2865841Z     452  452 |     } break;
2023-07-18T18:29:22.2866279Z     453  453 |     case PickleOpCode::APPEND: {
2023-07-18T18:29:22.2866863Z     453      |-      TORCH_CHECK(stack_.size() >= 2, "Parsing error: missing elements in stack_.");
2023-07-18T18:29:22.2867267Z          454 |+      TORCH_CHECK(
2023-07-18T18:29:22.2867684Z          455 |+          stack_.size() >= 2, "Parsing error: missing elements in stack_.");
2023-07-18T18:29:22.2868087Z     455  456 |       auto list_ivalue = stack_.at(stack_.size() - 2);
2023-07-18T18:29:22.2868474Z     456  457 |       readListElements(list_ivalue, stack_.size() - 1);
2023-07-18T18:29:22.2868786Z     457  458 |     } break;
2023-07-18T18:29:22.2869195Z --------------------------------------------------------------------------------
2023-07-18T18:29:22.2869513Z     647  648 | 
2023-07-18T18:29:22.2869851Z     648  649 |       auto dict = stack_.at(dict_pos).toGenericDict();
2023-07-18T18:29:22.2870360Z     649  650 |       dict.insert_or_assign(stack_.at(key_pos), stack_.at(val_pos));
2023-07-18T18:29:22.2870838Z     649      |-      stack_.erase(stack_.begin() + static_cast<ssize_t>(key_pos), stack_.end());
2023-07-18T18:29:22.2871232Z          651 |+      stack_.erase(
2023-07-18T18:29:22.2871627Z          652 |+          stack_.begin() + static_cast<ssize_t>(key_pos), stack_.end());
2023-07-18T18:29:22.2871955Z     651  653 |     } break;
2023-07-18T18:29:22.2872231Z     652  654 |     default: {
2023-07-18T18:29:22.2872509Z     653  655 |       AT_ERROR(
2023-07-18T18:29:22.2872901Z     653      |-          "Unknown opcode for unpickling: ",
2023-07-18T18:29:22.2873299Z     654      |-          int(static_cast<uint8_t>(opcode)));
2023-07-18T18:29:22.2873750Z          656 |+          "Unknown opcode for unpickling: ", int(static_cast<uint8_t>(opcode)));
2023-07-18T18:29:22.2874091Z     656  657 |     } break;
2023-07-18T18:29:22.2874334Z     657  658 |   }
2023-07-18T18:29:22.2874614Z     658  659 |   return opcode;
2023-07-18T18:29:22.2874819Z 
```

I got these by looking at https://hud.pytorch.org/pr/104027

ezyang commented:
Needs some tests!

ezyang reviewed (changes requested):
needs tests

emasap committed:
Add support for PickleOpCode::APPEND in torch unpickler (#104027)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/104027

Reviewed By: ezyang

Differential Revision: D46760650

fbshipit-source-id: e7e20c025af5860cbeb384c7ce95693dfd57c72e
